substrate/frame/delegated-staking/src/tests.rs

```diff
@@ -0,0 +1,647 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Tests for pallet-delegated-staking.
+
+use super::*;
+use crate::mock::*;
+use frame_support::{assert_noop, assert_ok, traits::fungible::InspectHold};
+use pallet_staking::Error as StakingError;
+use sp_staking::DelegationInterface;
+
+#[test]
+fn create_an_agent_with_first_delegator() {
+	ExtBuilder::default().build_and_execute(|| {
+		let agent: AccountId = 200;
+		let reward_account: AccountId = 201;
+		let delegator: AccountId = 202;
+
+		// set intention to accept delegation.
+		fund(&agent, 1000);
+		assert_ok!(DelegatedStaking::register_agent(
+			RawOrigin::Signed(agent).into(),
+			reward_account
+		));
+
+		// delegate to this account
+		fund(&delegator, 1000);
+		assert_ok!(DelegatedStaking::delegate_to_agent(
+			RawOrigin::Signed(delegator).into(),
+			agent,
+			100
+		));
+
+		// verify
+		assert!(DelegatedStaking::is_agent(&agent));
+		assert_eq!(DelegatedStaking::stakeable_balance(&agent), 100);
+		assert_eq!(
+			Balances::balance_on_hold(&HoldReason::StakingDelegation.into(), &delegator),
+			100
+		);
+		assert_eq!(DelegatedStaking::held_balance_of(&delegator), 100);
+	});
+}
+
+#[test]
+fn cannot_become_agent() {
+	ExtBuilder::default().build_and_execute(|| {
+		// cannot set reward account same as agent account
+		assert_noop!(
+			DelegatedStaking::register_agent(RawOrigin::Signed(100).into(), 100),
+			Error::<T>::InvalidRewardDestination
+		);
+
+		// an existing validator cannot become agent
+		assert_noop!(
+			DelegatedStaking::register_agent(
+				RawOrigin::Signed(mock::GENESIS_VALIDATOR).into(),
+				100
+			),
+			Error::<T>::AlreadyStaking
+		);
+
+		// an existing nominator cannot become agent
+		assert_noop!(
+			DelegatedStaking::register_agent(
+				RawOrigin::Signed(mock::GENESIS_NOMINATOR_ONE).into(),
+				100
```

Comment:
What if someone is staked but `Chill`-ed? can they become an agent? 

And in general, once you become an Agent, you are then exepcted to become a staker right?

Reply:
Anyone directly staked cannot become an agent (they can go through the migration path though.).

Yeah when an account become an agent, the pallet registers the account as a virtual staker in the staking pallet.

Reply:
Added some extra comments to clarify this.